src/features/SignAndVerifyMessage/SignMessage.tsx

```diff
@@ -80,106 +59,76 @@ const BackButton = styled(ButtonUI)<BackButtonProps>`
   }
 `;
 
-enum SignStatus {
-  NOT_SIGNED,
-  SIGNING,
-  SIGNED
-}
-
-interface SignProps {
+interface OwnProps {
   setShowSubtitle(show: boolean): void;
 }
 
-function SignMessage(props: Props) {
-  const [walletName, setWalletName] = useState<WalletId | undefined>(undefined);
-  const [wallet, setWallet] = useState<IFullWallet | null>(null);
-  const [signStatus, setSignStatus] = useState<SignStatus>(SignStatus.NOT_SIGNED);
-  const [message, setMessage] = useState('');
-  const [error, setError] = useState<string | undefined>(undefined);
-  const [signedMessage, setSignedMessage] = useState<ISignedMessage | null>(null);
-
-  const { setShowSubtitle, isDemoMode } = props;
-
-  const handleSignMessage = async () => {
-    setSignStatus(SignStatus.SIGNING);
-    try {
-      if (!wallet) {
-        throw Error;
-      }
-
-      const address = toChecksumAddress(wallet.getAddressString());
-      let sig = '';
-
-      let lib: Web3Node | undefined = undefined;
-      if (walletName && walletName !== WalletId.WALLETCONNECT && isWeb3Wallet(walletName)) {
-        lib = (await setupWeb3Node()).lib;
-      }
-      sig = await wallet.signMessage(message, lib);
-
-      const combined = {
-        address,
-        msg: message,
-        sig: addHexPrefix(sig),
-        version: '2'
-      };
-      setError(undefined);
-      setSignedMessage(combined);
-      setSignStatus(SignStatus.SIGNED);
-    } catch (err) {
-      setSignStatus(SignStatus.NOT_SIGNED);
-      setError(translateRaw('SIGN_MESSAGE_ERROR'));
-      console.debug('[SignMessage]', err);
+function SignMessage({
+  setShowSubtitle,
+  isDemoMode,
+  signedMessage,
+  message,
+  messageUpdate,
+  status,
+  error,
+  walletSelect,
+  walletId,
+  signMessageReset,
+  signMessage
+}: Props) {
+  const [wallet, setWallet] = useState<IFullWallet | undefined>(undefined);
+
+  useUnmount(() => {
+    // Kill WalletConnect session
+    if (wallet && walletId === WalletId.WALLETCONNECT) {
+      (wallet as WalletConnectWallet).kill();
     }
-  };
+    signMessageReset();
+  });
 
-  const handleOnChange = (msg: string) => {
-    setMessage(msg);
-    setError(undefined);
-    setSignStatus(SignStatus.NOT_SIGNED);
+  const onSelect = (walletId: WalletId) => {
+    setShowSubtitle(false);
+    walletSelect(walletId);
   };
 
-  const onSelect = (selectedWalletName: WalletId) => {
-    setWalletName(selectedWalletName);
-    setShowSubtitle(false);
+  const onUnlock = (w: IFullWallet) => {
+    const selectedWallet = Array.isArray(w) ? w[0] : w;
```

Comment:
In what case is `IFullWallet` an array?

Reply:
For example when a user authorises 2 accounts with MM.

Reply:
Can you explicitly write the type to allow for arrays then?